reducerfunction.py

```diff
@@ -0,0 +1,163 @@
+    file: Path,
+    force: bool = False,
+    jobs: int = multiprocessing.cpu_count(),
+) -> bool:
+    """Reduce a case given in the .tar format.
+    Interface for `reduced_code`.
+
+    Args:
+        file (Path): Path to .tar case.
+        force (bool): Force a reduction (even if the case is already reduced).
+    Returns:
+        bool: If the reduction was successful.
+    """
+    case = utils.Case.from_file(config, file)
+
+    if reduce_case(config=config, bldr=bldr, case=case, force=force, jobs=jobs):
+        case.to_file(file)
+        return True
+    return False
+
+
+def reduce_case(
+    config: utils.NestedNamespace,
+    bldr: ccbuilder.Builder,
+    case: utils.Case,
+    force: bool = False,
+    jobs: int = multiprocessing.cpu_count(),
+) -> bool:
+    """Reduce a case.
+
+    Args:
+        case (utils.Case): Case to reduce.
+        force (bool): Force a reduction (even if the case is already reduced).
+
+    Returns:
+        bool: If the reduction was successful.
+    """
+    if not force and case.reduced_code:
+
+        return True
+
+    case.reduced_code = reduce_code(
+        config=config,
+        bldr=bldr,
+        code=case.code,
+        marker=case.marker,
+        bad_setting=case.bad_setting,
+        good_settings=case.good_settings,
+        bisection=case.bisection,
+        jobs=jobs,
+    )
+    return bool(case.reduced_code)
+
+
+def reduce_code(
+    config: utils.NestedNamespace,
+    bldr: ccbuilder.Builder,
+    code: str,
+    marker: str,
+    bad_setting: utils.CompilerSetting,
+    good_settings: list[utils.CompilerSetting],
+    bisection: Optional[str] = None,
+    preprocess: bool = True,
+    jobs: Optional[int] = None,
+) -> Optional[str]:
+    """Reduce given code w.r.t. `marker`
+
+    Args:
+        code (str):
+        marker (str): Marker which exhibits the interesting behaviour.
+        bad_setting (utils.CompilerSetting): Setting which can not eliminate the marker.
+        good_settings (list[utils.CompilerSetting]): Settings which can eliminate the marker.
+        bisection (Optional[str]): if present the reducer will also check for the bisection
+        preprocess (bool): Whether or not to run the code through preprocessing.
+
+    Returns:
+        Optional[str]: Reduced code, if successful.
+    """
+
+    jobs = jobs if jobs else bldr.jobs
+
+    bad_settings = [bad_setting]
+    if bisection:
+        bad_settings.append(copy.deepcopy(bad_setting))
+        bad_settings[-1].rev = bisection
+        repo = bad_setting.repo
+        good_settings = good_settings + [copy.deepcopy(bad_setting)]
+        good_settings[-1].rev = repo.rev_to_commit(f"{bisection}~")
+
+    # creduce likes to kill unfinished processes with SIGKILL
+    # so they can't clean up after themselves.
+    # Setting a temporary temporary directory for creduce to be able to clean
+    # up everything
+    with diopter.utils.TempDirEnv() as tmpdir:
+        # preprocess file
+        if preprocess:
```

Comment:
Would it make sense to let the caller handle preprocessing?

Reply:
To a certain degree is already is the caller (wrt `diopter.reducer`) :D
But you are right.